### PR TITLE
Various fixes

### PR DESCRIPTION
**Pull request by rifeman2007**

### CODE REVIEW

File: src/AppBundle/Entity/User.php [hunk 1]
```diff
@@ -78,9 +78,9 @@ class User extends BaseUser
     /**
      * @var bool
      *
-     * @ORM\Column(type="boolean")
+     * @ORM\Column(type="boolean", nullable=true)
```
Comment (jongotlin) [inline src/AppBundle/Entity/User.php]: A boolean value should only have true or false as value.

**Comment by jongotlin** [comment]
Please also remove the Facebook javascript related code.

File: src/AppBundle/Entity/User.php (hunk 1 above)
Comment (rifeman2007) [inline src/AppBundle/Entity/User.php]: @jongotlin How would you handle now pre-selected value for want_to_learn field? It doesnt mean that boolean values can't be nullable.

Comment (rifeman2007) [inline src/AppBundle/Entity/User.php]: Your default value for $wantToLearn property is false which contradicts the task that I am assigned.

Task given:
Remove pre-set of field in sign up form (step 2 in the sign up)

Comment (jongotlin) [inline src/AppBundle/Entity/User.php]: Remove the default value like `protected $wantToLearn;`. The `ORM`-annotation is for database schema.

Comment (jongotlin) [inline src/AppBundle/Entity/User.php]: You might have to combine with `@Assert\NotBlank`.

Comment (rifeman2007) [inline src/AppBundle/Entity/User.php]: I did and it throws and exception in the database level. Do you really test this in your end?

Comment (jongotlin) [inline src/AppBundle/Entity/User.php]: What exception do you get?

**Comment by rifeman2007** [comment]
I think test fails in RegistrationControllerTest line 78. It expects 302 but it returned 200.

File: src/AppBundle/Entity/User.php (hunk 1 above)
Comment (jongotlin) [inline src/AppBundle/Entity/User.php]: If you do change the database schema you have to make an migration using Docitrine migrations. `php app/console doctrine:migrations:diff`. But as I said, not here.

**Comment by jongotlin** [comment]
Probably the form validation tells you the form is invalid due to missing want-to-learn value.

File: app/DoctrineMigrations/Version20160313161454.php [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+{
+    /**
+     * @param Schema $schema
+     */
+    public function up(Schema $schema)
+    {
+        // this up() migration is auto-generated, please modify it to your needs
+        $this->abortIf($this->connection->getDatabasePlatform()->getName() != 'mysql', 'Migration can only be executed safely on \'mysql\'.');
+
+        $this->addSql('ALTER TABLE fos_user CHANGE want_to_learn want_to_learn TINYINT(1) DEFAULT NULL');
```
Comment (jongotlin) [inline app/DoctrineMigrations/Version20160313161454.php]: Please don't allow a boolean value to be set to null.

Comment (rifeman2007) [inline app/DoctrineMigrations/Version20160313161454.php]: Please see screenshot http://screencast.com/t/iLviRlhU. The task assigned to me is not to allow pre-selected value for the $wantToLearn field. As I said it throws an exception if I don't do that.

Comment (jongotlin) [inline app/DoctrineMigrations/Version20160313161454.php]: A boolean value can only be either true or false. We have to find another solution than setting it to null. One is to move the user settings to a separate one-to-one-relation table.

Comment (rifeman2007) [inline app/DoctrineMigrations/Version20160313161454.php]: Just for your sake. Here is the error. 

SQLSTATE[23000]: Integrity constraint violation: 1048 Column 'want_to_learn' cannot be null

The steps to recreate the error.

1) Changed protected $wantToLearn = false to protected $wantToLearn
2) Signup for a new user.

Comment (rifeman2007) [inline app/DoctrineMigrations/Version20160313161454.php]: @jongotlin Kindly create a task for that and I will refactor it. Besides, UserType needs to be refactored because there is a query that is not supposed to be there. Not sure who worked on that knows about separation of concerns.

Comment (rifeman2007) [inline app/DoctrineMigrations/Version20160313161454.php]: @jongotlin You know what is really wrong with the $wantToLearn variable? It's how you assign it. Do you really think that it should be a boolean value? 

'user.form.want_to_learn.choice.learn'  => true,
 'user.form.want_to_learn.choice.teach'  => false,

Comment (jongotlin) [inline app/DoctrineMigrations/Version20160313161454.php]: No, it is a boolean value. The best solution is to move those settings to a user settings table. But in this case I think it's easier to set the data option. 
```php
'data' => $this->securityContext->hasRole('ROLE_USER_COMPLETED') ? $builder->getData()->getWantToLearn() : null
```
Not sure about the exact syntax but something like that. You have to inject the securityContext service by configuring the form type as a service.

Comment (jongotlin) [inline app/DoctrineMigrations/Version20160313161454.php]: Realised this form is used elsewhere too. Don't use the security context but check for role in `$builder->getData()`.

Comment (rifeman2007) [inline app/DoctrineMigrations/Version20160313161454.php]: @jongotlin OK, will try that approach.

**Comment by rifeman2007** [comment]
@jongotlin Also, kindly check the codes are some duplication. You might want to focus on that rather than small things.

**Comment by jongotlin** [comment]
Lets solve one thing in each pr. Can you put that code in a separate pr?

File: src/AppBundle/Controller/Admin/UserController.php [hunk 3]
```diff
@@ -38,8 +38,9 @@ public function viewAction(Request $request, User $user)
             new AdminUserType(),
             $user,
             [
-                'manager' => $this->getDoctrine()->getManager(),
-                'locale' => $request->getLocale(),
+                'manager'       => $this->getDoctrine()->getManager(),
+                'locale'        => $request->getLocale(),
+                'wantToLearn'   => (int)$user->getWantToLearn()
```
Comment (jongotlin) [inline src/AppBundle/Controller/Admin/UserController.php]: Don't cast to int. It's a boolean.

File: src/AppBundle/Controller/UserController.php [hunk 4]
```diff
@@ -26,8 +26,9 @@ public function settingsAction(Request $request)
             $user,
             [
                 'validation_groups' => ['settings'],
-                'manager' => $this->getDoctrine()->getManager(),
-                'locale' => $request->getLocale(),
+                'manager'           => $this->getDoctrine()->getManager(),
+                'locale'            => $request->getLocale(),
+                'wantToLearn'       => ($this->get('security.authorization_checker')->isGranted('ROLE_COMPLETE_USER') === false? null: (int)$user->getWantToLearn())
```
Comment (jongotlin) [inline src/AppBundle/Controller/UserController.php]: No casting

File: src/AppBundle/Controller/Admin/UserController.php (hunk 3 above)
Comment (rifeman2007) [inline src/AppBundle/Controller/Admin/UserController.php]: @jongotlin Passing false in data property does not pre-select the radio button. You are already over reacting on your code review. You should be focusing to make the code readable. Those functions are added in PHP be used. Can I see your code?

Comment (rifeman2007) [inline src/AppBundle/Controller/Admin/UserController.php]: You also might want to tell Symfony developers not to use casting https://github.com/symfony/symfony/blob/2.7/src/Symfony/Component/Form/Form.php#L348

File: app/Resources/FOSUserBundle/translations/validators.en.yml [hunk 5]
```diff
@@ -0,0 +1,26 @@
+fos_user:
```
Comment (jongotlin) [inline app/Resources/FOSUserBundle/translations/validators.en.yml]: Why did you add these two files?

File: src/AppBundle/Mailer/Mailer.php [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+namespace AppBundle\Mailer;
+
+use Symfony\Bundle\FrameworkBundle\Templating\EngineInterface;
+use Symfony\Component\Routing\RouterInterface;
+
+/**
+ * Class Mailer
+ * @package AppBundle\Mailer
+ */
+class Mailer
```
Comment (jongotlin) [inline src/AppBundle/Mailer/Mailer.php]: Can you combine this with the already existing `Service\EmailSender`?

File: src/AppBundle/Mailer/Mailer.php [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+     */
+    protected function sendEmailMessage($htmlRenderedTemplate, $txtRenderedTemplate, $subject, $toEmail, $fromEmail = 'info@kompisbyran.se', $replyEmail = 'matchning@kompisbyran.se')
+    {
+        $message = \Swift_Message::newInstance()
+            ->setSubject($subject)
+            ->setFrom($fromEmail)
+            ->setReplyTo($replyEmail)
+            ->setTo($toEmail)
+            ->setBody($htmlRenderedTemplate     , 'text/html')
+            ->addPart("$txtRenderedTemplate"    , 'text/plain')
```
Comment (jongotlin) [inline src/AppBundle/Mailer/Mailer.php]: Any reason why you put the variable in quotation?

File: src/AppBundle/Form/UserType.php [hunk 8]
```diff
@@ -33,19 +33,20 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
-                    true => 'user.form.want_to_learn.choice.learn',
-                    false => 'user.form.want_to_learn.choice.teach',
+                    'user.form.want_to_learn.choice.learn'  => '1',
+                    'user.form.want_to_learn.choice.teach'  => '0'
                 ],
-                'choice_value' => function ($currentChoiceKey) {
-                    return $currentChoiceKey ? 'true' : 'false';
-                }
+                'choices_as_values' => true,
+                'data'              => (!$user->hasRole('ROLE_COMPLETE_USER')? null: $user->getWantToLearn())
```
Comment (jongotlin) [inline src/AppBundle/Form/UserType.php]: :+1:

File: src/AppBundle/Controller/Admin/UserController.php (hunk 3 above)
Comment (jongotlin) [inline src/AppBundle/Controller/Admin/UserController.php]: Yes, and having a boolean variable to have the values true, false, 1, 0 or null was not readable. Looks much better now.

File: src/AppBundle/Mailer/Mailer.php (hunk 7 above)
Comment (rifeman2007) [inline src/AppBundle/Mailer/Mailer.php]: No reason at all. Just forgot to remove it.

File: src/AppBundle/Mailer/Mailer.php (hunk 6 above)
Comment (rifeman2007) [inline src/AppBundle/Mailer/Mailer.php]: Yes, I can combine but let us merge this PR first. Thanks

File: app/Resources/FOSUserBundle/translations/validators.en.yml (hunk 5 above)
Comment (rifeman2007) [inline app/Resources/FOSUserBundle/translations/validators.en.yml]: Those are intended for the validation message in FOSUserBundle. We can have in translation but it is much ideal to be there. And also that is how you override the translation of FOSUserBundle if you don't know it.

**Comment by rifeman2007** [comment]
@jongotlin Check this one https://github.com/kompisbyran/Kompisbyran/blob/master/src/AppBundle/Form/UserType.php#L19

and this duplication. Can't add any abstraction layer for that? Do you know DRY principles? https://github.com/kompisbyran/Kompisbyran/blob/master/src/AppBundle/Controller/UserController.php#L47

Also why are the controllers are FAT? 

These are the things that you should be looking for. You are just power tripping. :)

File: app/Resources/FOSUserBundle/translations/validators.en.yml (hunk 5 above)
Comment (jongotlin) [inline app/Resources/FOSUserBundle/translations/validators.en.yml]: I didn't see why you had to override them.

File: src/AppBundle/Mailer/Mailer.php (hunk 6 above)
Comment (jongotlin) [inline src/AppBundle/Mailer/Mailer.php]: Ok

**Comment by jongotlin** [comment]
You don't agree mixing data types is a bad thing? The app is far from perfect! I'm totally aware of that. It's built as a pro-bono project and hours are limited.

**Comment by jongotlin** [comment]
Even if there are some smells new code can be clean. Not mean to offend you but the datatype mixing was weird imo.

**Comment by mikaelohren** [comment]
I am really not looking for a perfect app. I am so thankful for your work Jon! Since our budget is limited on our hand also we can not really aim for a perfect app. Right now we have a mess with the system because of simple things such as email addresses being entered incorrectly and desperately need them solved.

**Comment by mikaelohren** [comment]
If this pr is mergeable I would really appreciate if we could do that.

**Comment by jongotlin** [comment]
Just want to know why the translation files are added. Then we can merge.
Next time @rifeman2007, please open separate prs.

**Comment by rifeman2007** [comment]
@jongotlin That is how you override FOSUserBundle files.

**Comment by jongotlin** [comment]
Yes, but why did you have to override them?

**Comment by rifeman2007** [comment]
I have added a message for the repeat email. Your form is also using the translation of FOSUserBundle.

**Comment by jongotlin** [comment]
Ah, now I see. But you don't have to override the whole file do you? Cant you just add
```yaml
fos_user:
    email:
        mismatch: The entered emails don't match
```

**Comment by mikaelohren** [comment]
Jon, are we ready for a merge now?

**Comment by jongotlin** [comment]
Merged and deployed! Thanks @rifeman2007 :clap:
